File: Data Structures & Algorithms/maximum-sum-circular-subarray/submission-1.rs
impl Solution {
    pub fn max_subarray_sum_circular(nums: Vec<i32>) -> i32 {
        let n = nums.len();
        if n == 0 {
            panic!("empty array");
        }

        let mut max_sum = nums[0];

        for right in (n - 1)..(2 * n - 1) {
            let left = right - (n - 1);

            let mut cur_sum = 0;
            for i in left..=right {
                let current = nums[i % n];
                cur_sum = cur_sum.max(0) + current;
                max_sum = max_sum.max(cur_sum);
            }
        }
        max_sum
    }
}

File: Data Structures & Algorithms/house-robber/submission-1.py
class Solution:
    def rob(self, nums: list[int]) -> int:
        dp = [0] * len(nums)

        def rob(idx: int) -> int:
            if idx >= len(nums):
                return 0
            dp[idx] = max(nums[idx] + rob(idx + 2), rob(idx + 1))
            return dp[idx]
        
        rob(0)
        return dp[0]